### ISSUE TITLE
ElasticFlow implementation tasks

### ISSUE THREAD

**Issue by zou000** [sql-machine-learning/elasticdl#16]
## ParameterServer
1. RPC interface
   - [ ] service definition 
   - [ ] SparseTensor support 

1. RPC implementation
   - [ ] TF/PyTorch tensor to.from Tensor proto bidirectional convertors
   - [ ] multi-threaded Server
   - [ ] Client 

1. PS implementation
   * Graph gradient update
   * large model support: model partition
   * large lookup table support (e.g. use redis)

## Data API
1. data fetcher: Need a unified way for feeding data. 
   * ODPS support for TF, PyTorch
      *  `tf.data.DataSet` and `torch.utils.data.Dataset`
   * Other data source?
1. data sharding
   * sharding methods
   * shard tracking:  re-queue shard when worker dies
      * use etcd?

## Worker with PS client/data fetcher/etcd client
1. PyTorch worker
1. Estimator Worker
   * Current POC crashes when iteration number is high, need to debug first
1. Other TF workers

## Master
* divide work, wait for workers to finish

## Job launcher
1. based on user code, decide PS and worker to launch
1. local launcher (for testing)
1. kubernetes launcher
   * on premise
   * AWS
   * GKE

## Elasticity, fault tolerance support
* periodic checkpoint/model saving (to file? redis?)
* initial model  loading (from file? redis? PS?)
* PS scaling
   * partitioned model support

## Kubernetes controller
1. resource tracking
1. scaling
